libvmi/driver/kvm/kvm_events.c

```diff
@@ -0,0 +1,987 @@
+            for (size_t msr_i=0; msr_i<msr_all_len; msr_i++) {
+                kvmi_reg = msr_index[msr_all[msr_i]];
+                if (kvm->libkvmi.kvmi_control_msr(kvm->kvmi_dom, vcpu, kvmi_reg, enabled)) {
+                    errprint("%s: failed to set MSR event for %s: %s\n", __func__,
+                             msr_to_str[msr_all[msr_i]], strerror(errno));
+                    continue;
+                }
+            }
+        }
+        dbprint(VMI_DEBUG_KVM, "--Set MSR events on all MSRs\n");
+    }
+
+    return VMI_SUCCESS;
+error_exit:
+    // disable monitoring
+    for (unsigned int vcpu = 0; vcpu < vmi->num_vcpus; vcpu++) {
+        if (KVMI_EVENT_CR == event_id)
+            kvm->libkvmi.kvmi_control_cr(kvm->kvmi_dom, vcpu, kvmi_reg, false);
+        if (KVMI_EVENT_MSR == event_id)
+            kvm->libkvmi.kvmi_control_msr(kvm->kvmi_dom, vcpu, kvmi_reg, false);
+    }
+    return VMI_FAILURE;
+}
+
+status_t
+kvm_set_intr_access(
+    vmi_instance_t vmi,
+    interrupt_event_t* event,
+    bool enabled)
+{
+#ifdef ENABLE_SAFETY_CHECKS
+    if (!vmi || !event)
+        return VMI_FAILURE;
+#endif
+
+    kvm_instance_t *kvm = kvm_get_instance(vmi);
+#ifdef ENABLE_SAFETY_CHECKS
+    if (!kvm || !kvm->kvmi_dom)
+        return VMI_FAILURE;
+#endif
+
+    switch (event->intr) {
+        case INT3:
+            for (unsigned int vcpu = 0; vcpu < vmi->num_vcpus; vcpu++)
+                if (kvm->libkvmi.kvmi_control_events(kvm->kvmi_dom, vcpu, KVMI_EVENT_BREAKPOINT, enabled)) {
+                    errprint("%s: failed to set event on VCPU %u: %s\n", __func__, vcpu, strerror(errno));
+                    goto error_exit;
+                }
+            break;
+        default:
+            errprint("KVM driver does not support enabling events for interrupt: %"PRIu32"\n", event->intr);
+            return VMI_FAILURE;
+    }
+
+    dbprint(VMI_DEBUG_KVM, "--%s interrupt %"PRIu32" monitoring\n",
+            (enabled) ? "Enabled" : "Disabled", event->intr);
+
+    return VMI_SUCCESS;
+error_exit:
+    // disable monitoring for all vcpus
+    for (unsigned int i = 0; i < vmi->num_vcpus; i++)
+        kvm->libkvmi.kvmi_control_events(kvm->kvmi_dom, i, KVMI_EVENT_BREAKPOINT, false);
+    return VMI_FAILURE;
+}
+
+status_t
+kvm_set_mem_access(
+    vmi_instance_t vmi,
+    addr_t gpfn,
+    vmi_mem_access_t page_access_flag,
+    uint16_t UNUSED(vmm_pagetable_id))
+{
+#ifdef ENABLE_SAFETY_CHECKS
+    if (!vmi) {
+        errprint("%s: invalid vmi handle\n", __func__);
+        return VMI_FAILURE;
+    }
+#endif
+    static bool pf_enabled = false;
+    unsigned char kvmi_access, kvmi_orig_access;
+    kvm_instance_t *kvm = kvm_get_instance(vmi);
+#ifdef ENABLE_SAFETY_CHECKS
+    if (!kvm || !kvm->kvmi_dom) {
+        errprint("%s: invalid kvm handle\n", __func__);
+        return VMI_FAILURE;
+    }
+#endif
+    // enable PF events the first time we call this function
```

Comment:
@mdontu I can't remember, can I move this at the driver initialization function `kvm_events_init` ?
Or does it have a performance cost ?

Reply:
We enable this event during initialization, because we always track memory access. So, we didn't measure it. But, as this setting won't trigger more VM-exists by itself I wouldn't worry that much.